### PR TITLE
PLANNER-1219 Fix strategic oscillation bug

### PR DESCRIPTION
**Pull request by triceo**
@ge0ffrey, I took the liberty of refreshing Jordan's PR, adding a test and cleaning up the code. This is ready for review and merge now.

### CODE REVIEW

**Comment by triceo** [comment]
Jenkins please retest this.

**Comment by triceo** [comment]
Build failed due to Sonar not being available. Jenkins please retest this.

**Comment by triceo** [comment]
Jenkins please retest this.

**Comment by ge0ffrey** [comment]
I need to go through this when I have some more time, I'll look at it next week.
Very good idea to add the unit test!

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/forager/finalist/StrategicOscillationByLevelFinalistPodium.java [hunk 1]
```diff
@@ -59,45 +63,77 @@ public void addMove(LocalSearchMoveScope moveScope) {
...
-            boolean moveIsHigher = ((Comparable) moveLevelNumbers[i]).compareTo(referenceLevelNumbers[i]) > 0;
-            boolean finalistIsHigher = ((Comparable) finalistLevelNumbers[i]).compareTo(referenceLevelNumbers[i]) > 0;
-            if (moveIsHigher) {
-                if (finalistIsHigher) {
-                    break;
-                } else {
-                    return 1;
+
+        Score moveScore = moveScope.getScore();
+        Score bestStepScore = moveScope.getStepScope().getScore();
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/forager/finalist/StrategicOscillationByLevelFinalistPodium.java]: This is always null at this point. The StepScope.getScore() gets filled in after the forager selects a winning move (which it gets from this podium).

**Comment by ge0ffrey** [comment]
Thanks @jordanovskijordan Great test! Sorry that I didn't pick on this sooner.

I merged it, but I edited the implementation. Here's my additional commit:
https://github.com/kiegroup/optaplanner/commit/55a769b6be668edd8e5aac2caa7a238f5086db65

Some of the old tests actually expected that improving moves weren't picked, so I had to adjust those - and adjust the implementation not to let those through.

**Comment by ge0ffrey** [comment]
@rsynek @michaltomco We need to validate the StrategicOscillationByLevelFinalistPodiumTest tests. There might be other bugs in here.

**Pull request merged**